README.md
# integration-reconciliation-report

Join two exported sides by a declared stable key, compare declared fields under
declared normalisation, and report **matched**, **missing**, **duplicated** and
**conflicting** records.

Two promises, because both are usually broken:

- **A duplicated key never overwrites itself.** Records are grouped, not
  indexed into a last-write-wins map. Two rows sharing a key produce a reported
  duplication — not a clean-looking match over whichever row was read last.
- **Nothing is rounded and no float is compared.** Amounts are compared as
  integer minor units at a declared precision; timestamps at a declared fixed
  offset and granularity. A value that does not fit the declaration is reported
  as unevaluated, never quietly made to fit.

The tool opens no socket and calls no provider. Two exported files and a plan
are the whole of the evidence.

- **Repository:** [edilec/integration-reconciliation-report](https://github.com/edilec/integration-reconciliation-report)
- **License:** MIT
- **Runtime:** Node 22 or later, no dependencies

## Install

```sh
npm install integration-reconciliation-report
```

Or run it from a checkout with no install step at all — the package has no
runtime and no development dependencies.

## Use

```sh
integration-reconciliation-report --root examples/clean
integration-reconciliation-report --root examples/broken --json | jq '.findings[].ruleId'
```

The JSON report goes to stdout and nothing else does, so stdout pipes straight
into a parser. The human summary and every diagnostic go to stderr; `--json`
suppresses the summary.

```
plan reconciliation.json: 3 field(s) compared under the declared normalisation.
source source.json: 6 record(s). destination destination.json: 5 record(s). 0 not indexed.
join: 6 key(s), 1 matched, 2 conflicting, 1 duplicated, 1 missing in destination, 1 missing in source, 0 unevaluated. status fail.
```

As a library:

```js
import { reconcileExports, exitCodeFor, formatReport } from 'integration-reconciliation-report'

const report = await reconcileExports({ root: 'examples/broken' })
process.stderr.write(formatReport(report))
process.exitCode = exitCodeFor(report)
```

### Exit codes

| Exit | Meaning | stdout |
| ---: | --- | --- |
| `0` | reconciled, and nothing broke | the report |
| `1` | reconciled, and at least one error-severity rule fired | the report |
| `2` | invalid configuration or bad usage | **empty** |
| `2` | evidence that could not be obtained; status `incomplete` | the report |

A consumer that pipes stdout must handle an empty stdout on exit 2: a
configuration error means the run never had a subject, so there is nothing to
report about. An input that could not be read *did* have a subject, so that run
emits an `incomplete` report naming the input it could not obtain.

## The three inputs

All three live inside one declared root, and all three must be distinct files.

`reconciliation.json` — the plan:

```json
{
  "schemaVersion": "1",
  "key": ["invoiceId"],
  "fields": [
    { "name": "amount", "type": "amount", "precision": 2, "currencyField": "currency" },
    { "name": "postedAt", "type": "date", "timezone": "+05:30", "granularity": "day" },
    { "name": "status", "type": "string", "trim": true, "caseSensitive": false }
  ]
}
```

`source.json` and `destination.json` — the two exported sides:

```json
{
  "schemaVersion": "1",
  "records": [
    { "invoiceId": "INV-1001", "amount": "1250.00", "currency": "INR",
      "postedAt": "2026-03-01T20:30:00Z", "status": "Posted" }
  ]
}
```

Properties the plan does not mention are data and are left alone. Unknown keys
in the plan or in an export *envelope* are refused by name: a one-character
typo that silently disabled a comparison would turn a real break into a green
run.

Field types are `amount`, `date`, `integer` and `string`. The full schema,
including which keys each type accepts, is in
[`docs/reconciliation-rules.md`](./docs/reconciliation-rules.md).

## What it reports

Each key present on either side gets exactly one outcome — `matched`,
`conflicting`, `duplicated`, `missingInDestination` or `missingInSource` — and
a fifth state is reported as itself rather than folded into one of them: a key
whose comparison needed evidence the export does not carry is **unevaluated**,
and makes the run `incomplete`.

Some rules worth knowing before the first run:

- `duplicate-key-in-source` / `duplicate-key-in-destination` name **every** row
  in the group. A duplicated key is not compared, because it has no single
  value to compare.
- `amount-not-exact` refuses a JSON number with a fractional part. `10.1` in a
  file is the binary double nearest 10.1; quote it and it compares exactly.
- `amount-exceeds-declared-precision` refuses a significant digit beyond the
  declared precision rather than rounding it away.
- `date-offset-missing` refuses a timestamp with no UTC offset, because it
  names a different instant for every reader.
- `plan-timezone-unsupported` refuses a named IANA zone. Only fixed offsets are
  supported, and an unsupported declaration makes the run incomplete rather
  than being silently treated as UTC.
- `key-case-collision` warns when two keys differ only by letter case — one
  system upper-casing its identifiers is a classic silent merge.
- `inputs-are-one-file` refuses two names for one inode. A hard link has no
  target, so a path comparison cannot see it, and a file joined against itself
  matches every key.

## Limits and non-goals

**What a `pass` means.** Every key present on either side was decided, and every
decided key agreed on every declared field at the declared precision, timezone
and granularity. That is all it means.

**What this tool cannot conclude:**

- It cannot show that two systems agree. It compares the fields the plan
  declares in the files it was given. A column neither export carries is not
  evidence, and a column the plan does not name is not compared.
- It cannot tell you whether the declared precision, timezone or granularity is
  the *right* one. Declaring `day` granularity makes two postings eleven hours
  apart equal; that is the declaration's doing, not a discovery.
- It cannot tell a missing record from a record that was never exported. Both
  sides are files. What the systems actually hold is not visible from here.
- It cannot decide which row of a duplicated group is authoritative. It reports
  the duplication and keeps every row; choosing between them is a business
  decision.
- It does no fuzzy matching, no similarity scoring and no transitive
  reconciliation. Two rows join when their declared key components are equal as
  text, and not otherwise.
- It does not support named IANA timezones. A named zone is reported as
  `plan-timezone-unsupported`, nothing is compared on that field and the run is
  `incomplete`. A leap second is a separate matter: `23:59:60` is not an
  instant this tool can place, so it is refused as `date-invalid` — a refusal
  of that one value, not support for the leap-second calendar.
- It knows no currency's minor unit. An amount is compared at the precision the
  plan declares and at no other, and `currencyField` only checks that both
  sides carry the same three-letter code. A currency whose smallest unit is not
  a power of ten — MGA and MRU are fifths — is compared at the declared
  precision like every other, and nothing reports it.
- It reads no database, calls no API and opens no socket. There is no "pull the
  live side" mode.
- It is read-only and writes nothing. There is no `--out`, no `--report` and no
  destination of any kind: `--plan`, `--source` and `--destination` all name
  *inputs*, and `--destination` is the second export to read, not a place to
  write. A run leaves its root byte-identical and creates nothing beside it.
  That matters because the output destination is where ten tools in this
  catalog destroyed a file they were never asked to touch; a tool with no
  destination cannot, and `test/destination.test.mjs` measures it through the
  real binary on every exit path rather than leaving it to be true by accident.

## Guarantees, and what holds them

- **A duplicated key is never resolved by overwrite.** Every record for a key is
  kept. `test/duplicates.test.mjs` is built so that both the keep-the-first and
  the keep-the-last variants fail it: each case asserts `duplicated`, `matched`
  and `conflicting` together, and asserts that a group of *n* rows names *n*
  pointers — which no map keyed by the join key can produce.
- **No float takes part in a comparison.** Amounts are `BigInt` minor units.
  Two amounts that are one binary double apart are still told apart, and a
  fractional JSON number is refused rather than compared.
- **Nothing is rounded silently.** An over-precise amount is refused with a
  finding; the run is incomplete, not a match and not a break.
- **The declared timezone decides.** The same instant in two zones is one
  bucket; two instants an hour apart are one day in `+05:30` and two days in
  `Z`, and the tests assert both.
- **`pass` is never reported on evidence that was not obtained**, and `pass`
  with `checked: 0` is not reachable. Every site that marks a run `incomplete`
  has a test asserting the status and the process exit code, so removing one
  turns exit 2 into exit 1 and fails.
- **Severity is pinned behaviourally.** One frozen `ruleId -> severity` table
  feeds every finding and an unknown rule id throws; the table is asserted
  against the documented catalog in both directions. Those are declarations, and
  a coordinated edit agrees with itself — so every rule is *also* pinned by
  running the real binary: the exit code where severity alone decides the
  verdict, and a literal error count plus the printed severity word where the
  run is incomplete either way. `test/severity-word.test.mjs` shares no map, no
  table and no builder with anything else.
- **Ordering is decided by code unit, and defended at every site.** A scan for
  `.localeCompare(` is not a determinism test — `Intl.Collator` collates
  identically and spells differently. `test/ordering.test.mjs` has a case for
  each of the eight places this tool orders something. **Six** are pinned with
  values the two comparators genuinely disagree about — `Z` against `a`, `a-b`
  against `a_b`, `INV-a` against `inv-a` — asserting the exact emitted
  sequence. The other **two** admit no such value: the list of known limit
  names in one diagnostic is lower camel case over `[A-Za-z]`, and the record
  listing inside a duplicated group is `/records/` followed by digits, and a
  plain collator cannot order either differently. Those two are reported as
  equivalent mutants and *proven* so — exhaustively, over every ordered pair of
  the eight real limit names and of the first forty-one record pointers — not
  claimed to be pinned. The drift that is reachable at the listing is a
  *numeric* collator, and the emitted sequence is pinned against one.
- **Every untrusted string is sanitised** — join keys, field names, paths,
  pointers, messages, suggestions and evidence alike — of C0, DEL, the whole C1
  range, U+2028, U+2029 and the bidi controls. A key carrying one of them is
  refused outright rather than cleaned up, and so is a key carrying whitespace
  the report would have to collapse — a tab, a doubled space, a no-break or
  ideographic space. A printed key is the key that was grouped, in full.
- **No input is quoted back, on the error path either.** A refused field value
  is described rather than echoed, and a file that does not parse is reported by
  position, line and column. `JSON.parse` embeds the input in one of its two
  error messages, so a file short enough to be only a credential would otherwise
  be reproduced in full by its own failure — and sanitising does not remove it,
  because the quotation is at the front of the message and the bound cuts from
  the back.
- **Every documented limit is enforced and tested from both sides**, and every
  flag is driven through the real binary so none can be documented and ignored.
- **Nothing opens a socket, reads a clock, reads a random source or reads the
  environment.**

## Verify

```sh
npm run check
```

Lint with `node --check` over every file, the `node:test` suite, the runnable
clean example, and a pack dry run. No network access and no install step.

## License

MIT. See [LICENSE](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project are documented in this file.

## Unreleased

### Fixed

- an input that does not parse is no longer quoted back. `JSON.parse` embeds
  the input in one of its two error messages
  (`Unexpected token 'A', "AKIA…" is not valid JSON`), so an export or a plan
  short enough to be only a credential was reproduced in full by
  `input-not-json` on stdout, in both output modes. That message was also the
  one untrusted string reaching a finding without passing through `excerpt`,
  so it carried control characters too. Both are fixed: `parseFailureDetail` in
  `src/text.mjs` keeps the position, line and column and drops the quotation,
  and the result is excerpted like everything else. Truncation alone would not
  have helped — the quotation is at the front of the message and the bound cuts
  from the back;

### Added

- `test/destination.test.mjs`, pinning the output destination this tool does not
  have. The destination is where ten tools in this catalog accepted a path that
  destroyed a file they were never asked to touch, four of them exiting 0; there
  is no `--out` here and nothing under `--root` is opened for writing, and both
  halves are now asserted through the real binary on every exit path -- the root
  byte-identical, nothing created beside it, every output-shaped flag refused as
  unknown with an empty stdout, and `--destination` proved to name an input to
  read rather than a file to create. Adding an unguarded `--out` to the binary,
  a sidecar inside the root, or a report beside the root each reddens the tests
  named for it. The source scan in `test/guarantees.test.mjs` now also names
  copy, link, symlink, cp, rm and the sync writers;
- a reconciliation plan — a declared join key of 1 to 8 components and a list
  of compared fields typed `amount`, `date`, `integer` or `string` — compiled
  from data, with every unknown key refused by name at both levels, including a
  key that is valid for a different field type, and a field that did not
  compile counted as uncompared rather than passed over;
- key grouping that **keeps every record**. Indexing into a last-write-wins map
  is the defining bug of a reconciliation tool: the second row for a key
  silently replaces the first, the group looks like a clean one-to-one match,
  and the duplicate is never reported. Every row is kept, the duplication is an
  error, and the group is not compared — a duplicated key has no single value
  to compare;
- amount comparison in integer minor units at a declared precision, on
  `BigInt`, so no floating-point value takes part. A JSON number with a
  fractional part is refused as the binary double it actually is; a significant
  digit beyond the declared precision is refused rather than rounded away;
  trailing zeros beyond the precision discard nothing and are accepted;
- date comparison as the bucket an instant falls in, at a declared fixed offset
  and granularity. The same instant written in two zones is one bucket at every
  granularity; two different instants are one day in `+05:30` and two days in
  `Z`. A timestamp with no offset is refused, a date with no time part is
  accepted only at `day` granularity, and the calendar is checked
  arithmetically, so `2026-02-29` is refused rather than rolled forward;
- named IANA timezones reported as **unsupported** rather than approximated:
  resolving one needs a tz database with daylight-saving history that this
  package does not carry, so the field is not compared, the rule fires and the
  run is `incomplete` — never silently treated as UTC;
- the four outcomes — matched, missing on either side, duplicated and
  conflicting — plus an explicit fifth state: a key whose comparison needed
  evidence the export does not carry is *unevaluated*, which is neither a match
  nor a break and makes the run incomplete;
- `amount-currency-conflict`, so two amounts are never declared equal across two
  currencies, and `key-case-collision`, so two keys that differ only by letter
  case are named before one system's upper-casing merges them;
- `field-match-after-normalization` and `duplicate-values-identical`, so a
  green run is not read as a claim that the two exports were spelled the same
  way, and a duplicated group that agrees with itself still says it is
  duplicated;
- `inputs-are-one-file`, which compares device and inode numbers rather than
  paths. `realpath` resolves a symbolic link, but a hard link has no target:
  two names for one inode are two distinct real paths, and a file joined
  against itself matches every key;
- strict UTF-8 decoding with `TextDecoder('utf-8', { fatal: true })` on all
  three inputs, the plan included — the configuration path is exactly where a
  sibling tool hardened its data path and forgot;
- real-path confinement on both sides for all three inputs, so a symbolic link
  planted inside the root is refused unread and a root reached through a link
  is not falsely refused; a name carrying `..` or an absolute path is refused
  earlier still, as a usage error with an empty stdout;
- explicit file-byte, record, key, per-key-group, field, key-component,
  field-length and finding limits, each reported by name when hit and each
  making the run `incomplete` instead of truncating. The key walk and the
  duplicate listing each name the first entry they did not reach;
- sanitisation of every untrusted string that reaches output — join keys, field
  names, paths, pointers, messages and suggestions as well as `evidence`, and
  an unknown CLI option on its way to stderr — covering C0 and DEL, the whole
  C1 range (U+0085 NEL forges a line of its own, U+009B is the 8-bit CSI),
  U+2028 and U+2029, and the bidi and isolate controls U+200E, U+200F,
  U+202A–U+202E and U+2066–U+2069, which are also refused inside a join key —
  as is any whitespace the rendering would collapse, so that a key component is
  printed verbatim and at its full length rather than as something four

test/cli.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'

import { AMOUNT, cliRun, fixture, planOf, row, sideOf, withRoot } from './support.mjs'

/** The command line: streams, exit codes, and every way a flag can be wrong. */

const CLEAN = fixture([row({ invoiceId: 'INV-1' })], [row({ invoiceId: 'INV-1' })])

test('--help and --version go to stdout and exit 0', async () => {
  for (const flag of ['--help', '-h']) {
    const run = await cliRun([flag])
    assert.equal(run.code, 0)
    assert.equal(run.stdout.startsWith('integration-reconciliation-report'), true)
    assert.equal(run.stdout.includes('--max-records-per-key'), true)
    assert.equal(run.stderr, '')
  }
  for (const flag of ['--version', '-v']) {
    const run = await cliRun([flag])
    assert.equal(run.code, 0)
    assert.equal(run.stdout.trim(), '0.1.0')
  }
})

test('the help text documents every flag the parser accepts', async () => {
  const { stdout } = await cliRun(['--help'])
  for (const flag of [
    '--root', '--plan', '--source', '--destination', '--json',
    '--max-file-bytes', '--max-records', '--max-keys', '--max-records-per-key',
    '--max-fields', '--max-key-fields', '--max-field-length', '--max-findings',
  ]) {
    assert.equal(stdout.includes(flag), true, `${flag} is not documented`)
  }
})

test('stdout carries the JSON report and nothing else', async () => {
  await withRoot(CLEAN, async (root) => {
    const run = await cliRun(['--root', root])
    assert.equal(run.code, 0)
    const report = JSON.parse(run.stdout)
    assert.equal(report.tool, 'integration-reconciliation-report')
    assert.equal(report.schemaVersion, '1')
    assert.equal(run.stdout.endsWith('}\n'), true)
    assert.equal(run.stderr.length > 0, true, 'a non-empty stderr is correct, not a failure')
    assert.equal(run.stderr.includes('{'), false, 'no part of the report leaks to stderr')
  })
})

test('--json suppresses the human summary and nothing else', async () => {
  await withRoot(CLEAN, async (root) => {
    const plain = await cliRun(['--root', root])
    const quiet = await cliRun(['--root', root, '--json'])
    assert.equal(plain.stdout, quiet.stdout)
    assert.equal(quiet.stderr, '')
    assert.equal(plain.stderr.includes('status pass'), true)
  })
})

test('--plan, --source and --destination are honoured', async () => {
  const files = {
    'plan.json': planOf([AMOUNT]),
    'left.json': sideOf([row({ invoiceId: 'INV-1' })]),
    'right.json': sideOf([row({ invoiceId: 'INV-1' })]),
  }
  await withRoot(files, async (root) => {
    const run = await cliRun(['--root', root, '--plan', 'plan.json', '--source', 'left.json', '--destination', 'right.json', '--json'])
    assert.equal(run.code, 0)
    assert.equal(JSON.parse(run.stdout).summary.matched, 1)

    const stderrRun = await cliRun(['--root', root, '--plan', 'plan.json', '--source', 'left.json', '--destination', 'right.json'])
    assert.equal(stderrRun.stderr.includes('plan plan.json'), true)
    assert.equal(stderrRun.stderr.includes('source left.json'), true)
    assert.equal(stderrRun.stderr.includes('destination right.json'), true)
  })
})

test('a configuration error leaves stdout empty and exits 2', async () => {
  for (const args of [[], ['--root'], ['--root', '.', '--nope'], ['--root', '.', '--source']]) {
    const run = await cliRun(args)
    assert.equal(run.code, 2, JSON.stringify(args))
    assert.equal(run.stdout, '', JSON.stringify(args))
    assert.equal(run.stderr.length > 0, true)
  }
})

test('a value-carrying flag given twice is refused rather than silently last-wins', async () => {
  await withRoot(CLEAN, async (root) => {
    const run = await cliRun(['--root', root, '--source', 'a.json', '--source', 'source.json'])
    assert.equal(run.code, 2)
    assert.equal(run.stdout, '')
    assert.equal(run.stderr.startsWith('--source was given more than once'), true)

    const limit = await cliRun(['--root', root, '--max-records', '5', '--max-records', '50000'])
    assert.equal(limit.code, 2)
    assert.equal(limit.stdout, '')
  })
})

test('--json may repeat, because it carries no value to lose', async () => {
  await withRoot(CLEAN, async (root) => {
    const run = await cliRun(['--root', root, '--json', '--json'])
    assert.equal(run.code, 0)
    assert.equal(run.stderr, '')
  })
})

test('an incomplete run says so on stderr as well as in the report', async () => {
  const files = fixture([row({ invoiceId: 'INV-0', amount: '1.00' }), { amount: '2.00' }], [row({ invoiceId: 'INV-0', amount: '1.00' })])
  await withRoot(files, async (root) => {
    const run = await cliRun(['--root', root])
    assert.equal(run.code, 2)
    assert.equal(JSON.parse(run.stdout).status, 'incomplete')
    assert.equal(run.stderr.includes('this run is not a pass.'), true)
  })
})

test('the three exit codes are reachable from the command line', async () => {
  await withRoot(CLEAN, async (root) => {
    assert.equal((await cliRun(['--root', root, '--json'])).code, 0)
  })
  await withRoot(fixture([row({ invoiceId: 'INV-1' })], []), async (root) => {
    assert.equal((await cliRun(['--root', root, '--json'])).code, 1)
  })
  await withRoot(fixture([], []), async (root) => {
    assert.equal((await cliRun(['--root', root, '--json'])).code, 2)
  })
})


src/plan.mjs
/**
 * The reconciliation plan: which fields form the join key, and how each
 * compared field is normalised before two spellings are called equal.
 *
 * The plan is configuration that happens to live in a file. Nothing in it is
 * inferred: a precision, a timezone and a granularity are declared or the
 * field is refused, because guessing any of the three decides matches and
 * breaks on the reader's behalf. Every unknown key is refused rather than
 * ignored -- a one-character typo that disables a comparison would turn a real
 * break into a green run, which is the defect this whole catalog is built
 * against.
 */

import { GRANULARITIES, MAX_PRECISION, MIN_PRECISION, parseTimezone } from './normalize.mjs'
import { describeValue, excerpt, isPlainObject } from './text.mjs'

export const PLAN_SCHEMA_VERSION = '1'

/** Top-level plan keys. Anything else is refused by name. */
export const PLAN_KEYS = Object.freeze(['fields', 'key', 'schemaVersion'])

/** The field types this tool knows how to normalise. */
export const FIELD_TYPES = Object.freeze(['amount', 'date', 'integer', 'string'])

/** Which keys each field type accepts. A key that belongs to another type is refused here. */
export const FIELD_KEYS = Object.freeze({
  amount: Object.freeze(['currencyField', 'name', 'precision', 'type']),
  date: Object.freeze(['granularity', 'name', 'timezone', 'type']),
  integer: Object.freeze(['name', 'type']),
  string: Object.freeze(['caseSensitive', 'collapseWhitespace', 'name', 'trim', 'type']),
})

const FIELD_NAME = /^[A-Za-z0-9][A-Za-z0-9_.-]{0,63}$/

/** A field name is a literal object key in a record. It is never a path: a dot means a dot. */
export function isFieldName(value) {
  return typeof value === 'string' && FIELD_NAME.test(value)
}

function optionalBoolean(sink, file, pointer, value, fallback, what) {
  if (value === undefined) return fallback
  if (typeof value !== 'boolean') {
    sink.add({
      file,
      pointer,
      ruleId: 'plan-field-invalid',
      message: `${what} must be true or false; the plan declares ${describeValue(value)}.`,
    })
    return null
  }
  return value
}

function compileField(sink, file, index, raw, keyFields) {
  const pointer = `/fields/${index}`
  if (!isPlainObject(raw)) {
    sink.add({
      file,
      pointer,
      ruleId: 'plan-field-invalid',
      message: `Field ${index} must be an object; the plan declares ${describeValue(raw)}.`,
    })
    return null
  }
  if (!isFieldName(raw.name)) {
    sink.add({
      file,
      pointer: `${pointer}/name`,
      ruleId: 'plan-field-invalid',
      message: 'A field name must be 1-64 characters of letters, digits, ".", "-" or "_", starting with a letter or digit.',
      suggestion: 'Name the record property exactly as it is spelled in the export; a field name is a literal key, never a path.',
    })
    return null
  }
  const name = raw.name
  if (typeof raw.type !== 'string' || !FIELD_TYPES.includes(raw.type)) {
    sink.add({
      file,
      pointer: `${pointer}/type`,
      ruleId: 'plan-type-unsupported',
      message: `Field "${excerpt(name, 64)}" declares a type this tool does not support; the supported types are ${FIELD_TYPES.join(', ')}.`,
      suggestion: 'An unsupported type is reported rather than compared, because comparing it would mean guessing how to normalise it.',
    })
    return null
  }
  const type = raw.type

  let refused = false
  for (const key of Object.keys(raw)) {
    if (FIELD_KEYS[type].includes(key)) continue
    refused = true
    sink.add({
      file,
      pointer: `${pointer}/${key}`,
      ruleId: 'plan-field-invalid',
      message: `Field "${excerpt(name, 64)}" of type ${type} does not accept the key "${excerpt(key, 64)}"; it accepts ${FIELD_KEYS[type].join(', ')}.`,
      suggestion: 'An unknown key is refused rather than ignored: a typo that silently disabled a comparison would turn a real break into a pass.',
    })
  }
  if (keyFields.includes(name)) {
    refused = true
    sink.add({
      file,
      pointer: `${pointer}/name`,
      ruleId: 'plan-field-invalid',
      message: `Field "${excerpt(name, 64)}" is also a join key field, so both sides hold the same value by construction and comparing it proves nothing.`,
      suggestion: 'Remove it from "fields", or join on a different key.',
    })
  }
  if (refused) return null

  const spec = { name, type, pointer }

  if (type === 'amount') {
    if (!Number.isInteger(raw.precision) || raw.precision < MIN_PRECISION || raw.precision > MAX_PRECISION) {
      sink.add({
        file,
        pointer: `${pointer}/precision`,
        ruleId: 'plan-field-invalid',
        message: `Amount field "${excerpt(name, 64)}" must declare an integer precision between ${MIN_PRECISION} and ${MAX_PRECISION}; the plan declares ${describeValue(raw.precision)}.`,
        suggestion: 'Precision is the number of decimal places the two systems agree to compare at. It is declared, never inferred.',
      })
      return null
    }
    spec.precision = raw.precision
    if (raw.currencyField !== undefined) {
      if (!isFieldName(raw.currencyField) || raw.currencyField === name) {
        sink.add({
          file,
          pointer: `${pointer}/currencyField`,
          ruleId: 'plan-field-invalid',
          message: `Amount field "${excerpt(name, 64)}" declares a currencyField that is not a distinct field name.`,
        })
        return null
      }
      spec.currencyField = raw.currencyField
    }
    return spec
  }

  if (type === 'date') {
    const zone = parseTimezone(raw.timezone)
    if (!zone.ok) {
      sink.add({
        file,
        pointer: `${pointer}/timezone`,
        ruleId: 'plan-timezone-unsupported',
        message:
          `Date field "${excerpt(name, 64)}" declares a timezone this tool does not support. ` +
          'Only a fixed offset -- "Z", "+HH:MM" or "-HH:MM" -- is supported; a named zone needs a tz database with daylight-saving history, which this package does not carry.',
        suggestion: 'Declare the fixed offset the two systems reconcile in, for example "+05:30".',
      })
      return null
    }
    if (typeof raw.granularity !== 'string' || !GRANULARITIES.includes(raw.granularity)) {
      sink.add({
        file,
        pointer: `${pointer}/granularity`,
        ruleId: 'plan-granularity-unsupported',
        message: `Date field "${excerpt(name, 64)}" must declare one of ${GRANULARITIES.join(', ')} as its granularity.`,
        suggestion: 'Granularity says how close two instants must be to count as the same posting. It is declared, never inferred.',
      })
      return null
    }
    spec.offsetMinutes = zone.minutes
    spec.timezone = zone.label
    spec.granularity = raw.granularity
    return spec
  }

  if (type === 'string') {
    const trim = optionalBoolean(sink, file, `${pointer}/trim`, raw.trim, true, `Field "${excerpt(name, 64)}" key "trim"`)
    const caseSensitive = optionalBoolean(sink, file, `${pointer}/caseSensitive`, raw.caseSensitive, true, `Field "${excerpt(name, 64)}" key "caseSensitive"`)
    const collapse = optionalBoolean(sink, file, `${pointer}/collapseWhitespace`, raw.collapseWhitespace, false, `Field "${excerpt(name, 64)}" key "collapseWhitespace"`)
    if (trim === null || caseSensitive === null || collapse === null) return null
    spec.trim = trim
    spec.caseSensitive = caseSensitive
    spec.collapseWhitespace = collapse
    return spec
  }

  return spec
}

/**
 * Compile the plan document.
 *
 * Returns `null` when nothing usable survived. A plan that compiled with one
 * field refused returns the rest together with `declaredFields`, so the caller
 * can see that fewer fields were compared than the plan declares and mark the
 * run incomplete rather than reporting a pass over a comparison that never ran.
 */
export function compilePlan(sink, file, document, limits) {
  if (!isPlainObject(document)) {
    sink.add({
      file,
      pointer: '',
      ruleId: 'plan-invalid',
      message: `The plan must be a JSON object; ${file} holds ${describeValue(document)}.`,
    })
    return null
  }
  let fatal = false
  for (const key of Object.keys(document)) {
    if (PLAN_KEYS.includes(key)) continue
    fatal = true
    sink.add({
      file,
      pointer: `/${key}`,
      ruleId: 'plan-key-unknown',
      message: `The plan declares an unknown key "${excerpt(key, 64)}"; it accepts ${PLAN_KEYS.join(', ')}.`,
      suggestion: 'An unknown key is refused rather than ignored: a typo that silently disabled part of the plan would turn a real break into a pass.',
    })
  }
  if (document.schemaVersion !== PLAN_SCHEMA_VERSION) {
    fatal = true
    sink.add({
      file,
      pointer: '/schemaVersion',
      ruleId: 'plan-invalid',
      message: `The plan must declare schemaVersion "${PLAN_SCHEMA_VERSION}"; it declares ${describeValue(document.schemaVersion)}.`,
    })
  }

  if (!Array.isArray(document.key) || document.key.length === 0) {
    sink.add({
      file,
      pointer: '/key',
      ruleId: 'plan-invalid',
      message: `The plan must declare "key" as a non-empty array of field names; it declares ${describeValue(document.key)}.`,
      suggestion: 'The join key is the stable identifier both systems export. Without one there is nothing to join on.',
    })
    return null
  }
  if (document.key.length > limits.maxKeyFields) {
    sink.add({
      file,
      pointer: '/key',
      ruleId: 'too-many-key-fields',
      message: `The plan declares ${document.key.length} key field(s), above the maxKeyFields limit of ${limits.maxKeyFields}; nothing was joined.`,
      suggestion: 'Raise --max-key-fields, or join on fewer fields.',
    })
    return null
  }
  const keyFields = []
  for (let index = 0; index < document.key.length; index += 1) {
    const name = document.key[index]
    if (!isFieldName(name)) {
      sink.add({
        file,
        pointer: `/key/${index}`,
        ruleId: 'plan-invalid',
        message: `Key field ${index} must be 1-64 characters of letters, digits, ".", "-" or "_", starting with a letter or digit.`,
      })
      return null
    }
    if (keyFields.includes(name)) {
      sink.add({
        file,
        pointer: `/key/${index}`,
        ruleId: 'plan-field-duplicate',
        message: `Key field "${excerpt(name, 64)}" is declared more than once.`,
      })
      return null
    }
    keyFields.push(name)
  }

  if (!Array.isArray(document.fields)) {
    sink.add({
      file,
      pointer: '/fields',
      ruleId: 'plan-invalid',
      message: `The plan must declare "fields" as an array; it declares ${describeValue(document.fields)}.`,
    })
    return null
  }
  if (document.fields.length === 0) {
    sink.add({
      file,
      pointer: '/fields',
      ruleId: 'no-fields-declared',
      message: 'The plan declares no fields to compare, so a matched key would be green on no field evidence at all.',
      suggestion: 'Declare at least one field to compare, even if it is only the amount.',
    })
    return null
  }
  if (document.fields.length > limits.maxFields) {
    sink.add({
      file,
      pointer: '/fields',
      ruleId: 'too-many-fields',
      message: `The plan declares ${document.fields.length} field(s), above the maxFields limit of ${limits.maxFields}; nothing was compared.`,
      suggestion: 'Raise --max-fields, or compare fewer fields.',
    })
    return null
  }

  const fields = []
  const seen = new Set()
  for (let index = 0; index < document.fields.length; index += 1) {
    const spec = compileField(sink, file, index, document.fields[index], keyFields)
    if (spec === null) continue
    if (seen.has(spec.name)) {
      sink.add({
        file,
        pointer: `/fields/${index}/name`,
        ruleId: 'plan-field-duplicate',
        message: `Field "${excerpt(spec.name, 64)}" is declared more than once; the second declaration was refused rather than silently replacing the first.`,
      })
      continue
    }
    seen.add(spec.name)
    fields.push(spec)
  }

  if (fatal || fields.length === 0) return null
  return { key: keyFields, fields, declaredFields: document.fields.length }
}
